contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,317 @@
+
+    function setContractClosed(address target) external;
+
+    // TODO: Ability to eject a contract. See AccessManaged.setAuthority
+    // function transferContractAuthority(address target) external;
+}
+
+/**
+ * @dev AccessManager is a central contract that stores the permissions of a system. It is an AccessControl contract,
+ * i.e. it has roles and all the standard functions like `grantRole` and `revokeRole`, but it defines a particular set
+ * of roles, with a particular structure.
+ *
+ * Users are grouped in "teams". Teams must be created before users can be assigned into them, up to a maximum of 255
+ * teams. A user can be assigned to multiple teams. Each team defines an AccessControl role, identified by a role id
+ * that starts with the ASCII characters `team:`, followed by zeroes, and ending with the single byte corresponding to
+ * the team number.
+ *
+ * Contracts in the system are grouped as well. These are simply called "contract groups". There can be an arbitrary
+ * number of groups. Each contract can only be in one group at a time. In the simplest setup, each contract is assigned
+ * to its own separate group, but groups can also be shared among similar contracts.
+ *
+ * All contracts in a group share the same permissioning scheme. A permissioning scheme consists of a mapping between
+ * functions and allowed teams. Each function can be allowed to multiple teams, meaning that if a user is in at least
+ * one of the allowed teams they can call that function.
+ *
+ * Note that a function in a target contract may become permissioned only when: 1) said contract is {AccessManaged} and
+ * is connected to this contract as its manager, and 2) said function is decorated with the `restricted` modifier.
+ *
+ * There is a special team defined by default named "all" of which all accounts are considered members, and two special
+ * contract groups: 1) the "open" team, where all functions are allowed to the "all" team, and 2) the "closed" team,
+ * where no function is allowed to any team.
+ *
+ * Permissioning schemes and team and contract group assignments can be configured by the default admin. The contract
+ * includes {AccessControlDefaultAdminRules} by default to enforce security rules on this account. Additionally, it is
+ * expected that the account will be highly secured (e.g., a multisig or a well-configured DAO) as all the permissions
+ * of the managed system can be modified by it.
+ */
+contract AccessManager is IAccessManager, AccessControlDefaultAdminRules {
+    bytes32 _createdTeams;
+    mapping(address => bytes32) private _userTeams;
+    mapping(bytes32 => mapping(bytes4 => bytes32)) private _allowedTeams;
+    mapping(address => bytes32) private _contractGroup;
+
+    uint8 private constant _TEAM_ALL = 255;
+    bytes32 private constant _GROUP_UNSET = 0;
+    bytes32 private constant _GROUP_OPEN = "group:open";
+    bytes32 private constant _GROUP_CLOSED = "group:closed";
```

Comment:
I would have gone for hashes like we do everywhere else. Sure that syntax is nice because you can understand the value without doing a preimage, but that might brings confusion as to why we use it here and not for roles such as PROPOSER_ROLE (in the timelock).

Reply:
In retrospect I think it was a mistake to use hashes and I think this is a clearly superior option. I think you're understating the benefit of not needing the preimage!

Reply:
The improvement that I see is that you are able to get a value (role id, group id) and more easily interpret it. With hashes, it's very difficult to map a value to its intended "meaning", you need a rainbow table to get an actual label that explains that meaning in a human friendly way.

That said, this benefit is not fully realized because standard tooling will automatically display a `bytes32` value as a hex string. But I still think it's worth doing without hashes for applications that integrate with AccessManager. For example, if we add a new special group in the future called `group:paused` or something, those applications would be automatically able to show `group:paused` in the UI without needing an update to the rainbow table.